BeeStation/BeeStation-Hornet
# Updates dismemberment/delimbing and weapon sharpness

## About The Pull Request

This adds two new categories of sharpness on top of the existing two, all four sharpness categories are:
* `IS_BLUNT` which is for blunt weapons that are not sharp.
* `IS_SHARP` which is for sharp, but smaller or lighter objects such as knives and scalpels.
* `IS_SHARP_DISMEMBER` is for weapons such as swords, axes and chainsaws, and also includes the chef's cleaver and due to its smaller size the energy dagger. 
* `IS_SHARP_DISMEMBER_EASY` is primarily for weapons with energy blades such as energy swords and the ninja's energy katana, the weapons that cut through flesh and bone like butter. The nightmare's light eater is the only non-energy weapon that fits in this category.

In addition to the above this PR refactors how dismemberment is calculated entirely:
* If the mob has trait for easy limb dismemberment (such as skeletons/oozelings) **and** the limb has already sustained at least some damage, any weapons (even blunt ones) have a chance equal to their damage output to dismember it.
* If the weapon is sharp enough for easy delimbing, it has a chance equal to its damage output for immediate dismemberment, even if the limb has taken no prior damage.
* If the weapon is sharp enough for normal dismemberment, it will only dismember a limb that has already been fully disabled with brute/burn damage before the current attack. Fire axes are cleaving chunks out of people, but they won't fully cut a limb off without at least three chops.
* If the weapon is sharp, but not sharp enough for normal dismemberment, it can only dismember if the target is dead.
* If the weapon is blunt, the only way it can dismember is meeting the criteria in the first bullet point. 
* If the limb being targetted is the head, even if *other* criteria listed above is met, the dismember will fail if the target is not at least in crit. No beheading people who are still upright for game balance reasons, even if an energy sword would/should be able to do so. 
* If the mob can't be delimbed for whatever reason, all of the above will fail to delimb

Now for some examples from each category:
* An energy sword has `IS_SHARP_DISMEMBER_EASY` so every swing will have a 30% chance to dismember the targetted limb. Because it does 30 damage, even if the target is armored the limb will probably be disabled in two hits, meaning the third hit is guaranteed to dismember the targetted limb. 
* A fire axe has `IS_SHARP_DISMEMBER` so it can only dismember a targetted limb that has already been fully disabled. This means against an unarmored person it will take a minimum of three hits, and some armored players will take four. 
* A laser scalpel has `IS_SHARP` so it can only delimb a player who is already completely dead
* A toolbox can never delimb anyone unless the target has the easy delimb trait themselves. Then it has a 12% chance per attack to delimb



## Why It's Good For The Game

In #11802 I brought up the testing done in the battle royal server's environment, but there was a critical change that affected delimbing on Bee since then which I didn't consider. The aftermath of that PR and the fact that item weights are factored into delimbing made for several unusual cases.

After discussion with Bacon, this seems to address most of the balance issues with delimbing in general, while still keeping with the core spirit of the PR and improved game balance overall. Security still has to deal with potentially being de-limbed, and all players still have more reason to target specific limbs depending on their exact priorities. 

If it is decided that limb targeting is bad for the game overall, this PR should simply be closed and #11802 fully reverted.



## Testing Photographs and Procedure

<details>
<summary>Testing evidence here</summary>

![image](https://github.com/user-attachments/assets/13a7fdbc-d284-446d-b5b4-a98842ab9556)

![image](https://github.com/user-attachments/assets/bd1a0e49-39b4-44f7-b8e6-594f4272c2ca)

Still pictures don't really do much for this, so here's the exact breakdown of testing:
* Spawn blunt object and skeleton. Whack skeleton until arm comes off
* Spawn normal human, whack leg until death to see if it ever comes off
* Spawn normal human and energy Katana, observe random delimbing as well as guaranteed delimbing in three hits
* Spawn normal human and fire axe, observe exlusively three hit delimbing, repeat three times
* Spawn human and scalpel, kill them through limb damage and observe dismemberment immediately after death
* Change force value on Katana to 100 and observe 100% delimbing chance
* Realize I can decapitate instantly with energy weapons and fix the code for it
* Verify decapitation is no longer possible on fully conscious targets

</details>

## Changelog
:cl:
balance: Completely overhauled dismemberment and weapon sharpness. Weapons now have four levels of sharpness: Blunt, Sharp, Dismember and Easy Dismember. 
balance: Blunt weapons are now only able to cause dismemberment on species that have the easy dismemberment trait, all of which can easily regenerate or otherwise slap their dismembered limb back into place.
balance: Species with the easy dismemberment trait (Skeletons, IPCs, Oozelings) have a chance equal to the damage received to lose a targetted limb as long as the limb was already damaged before being struck. 
balance: Ordinary sharp weapons such as knives and surgical tools can only dismember a dead target. 
balance: Large weapons such as swords, axes and chainsaws are only able to dismember a limb which has already been fully disabled. This means that it will take at least three hits with a fire axe to take someone's leg off. 
balance: Bladed energy weapons have a chance equal to their damage to dismember a targetted limb regardless of how much damage it has taken, in addition to being guaranteed to dismember a limb that has been fully disabled by damage. The nightmare's light eater is the only non-energy weapon in this group right now. 
balance: Mobs cannot be decapitated through the simplified mechanics while they are fully conscious - they must at minimum be asleep or in crit in order for dismemberment to apply when targetting someone's head for balance reasons.
/:cl:
